uberfire-extensions/uberfire-metadata/uberfire-metadata-backends/uberfire-metadata-backend-infinispan/src/main/java/org/uberfire/ext/metadata/backend/infinispan/proto/KObjectMarshaller.java

```diff
@@ -0,0 +1,287 @@
+                                            CLUSTER_ID,
+                                            SEGMENT_ID,
+                                            MetaObject.META_OBJECT_KEY,
+                                            MetaObject.META_OBJECT_FULL_TEXT);
+    }
+
+    @Override
+    public KObjectImpl readFrom(ProtoStreamReader protoStreamReader) throws IOException {
+
+        Descriptor descriptor = protoStreamReader.getSerializationContext().getMessageDescriptor(this.getTypeName());
+
+        List<KProperty<?>> properties = descriptor.getFields()
+                .stream()
+                .filter(fieldDescriptor -> !isMainAttribute(fieldDescriptor))
+                .filter(fieldDescriptor ->
+                                isExtension(descriptor.getName())
+                )
+                .map(field -> (KProperty<?>) new KPropertyImpl(toKPropertyFormat(field.getName()),
+                                                               this.read(field,
+                                                                         protoStreamReader),
+                                                               false)
+                )
+                .collect(toList());
+
+        String id = protoStreamReader.readString(MetaObject.META_OBJECT_ID);
+        String type = protoStreamReader.readString(MetaObject.META_OBJECT_TYPE);
+        String clusterId = protoStreamReader.readString(CLUSTER_ID);
+        String segmentId = protoStreamReader.readString(SEGMENT_ID);
+        String key = protoStreamReader.readString(MetaObject.META_OBJECT_KEY);
+        String fullText = protoStreamReader.readString(MetaObject.META_OBJECT_FULL_TEXT);
+
+        return new KObjectImpl(id,
+                               type,
+                               clusterId,
+                               segmentId,
+                               key,
+                               properties,
+                               !StringUtils.isEmpty(fullText));
+    }
+
+    @Override
+    public void writeTo(ProtoStreamWriter protoStreamWriter,
+                        KObject kObject) throws IOException {
+
+        protoStreamWriter.writeString(MetaObject.META_OBJECT_ID,
+                                      kObject.getId());
+        protoStreamWriter.writeString(MetaObject.META_OBJECT_TYPE,
+                                      kObject.getType().getName());
+        protoStreamWriter.writeString(CLUSTER_ID,
+                                      kObject.getClusterId());
+        protoStreamWriter.writeString(SEGMENT_ID,
+                                      kObject.getSegmentId());
+        protoStreamWriter.writeString(MetaObject.META_OBJECT_KEY,
+                                      kObject.getKey());
+
+        List<KProperty<?>> props = new ArrayList<>();
+        kObject.getProperties().iterator().forEachRemaining(props::add);
+
+        props.removeIf(kProperty -> props.stream()
```

Comment:
This looks rather inefficient. For each property we're checking all the other properties and counting how many of them have the same name. This algorithm has quadratic complexity. This means if you have KObject with 10 uniquely named properties it will have to perform roughly 100 comparisons.

Why is this check even needed? Are there KObjects with multiple fields with the same name?

A more efficient solution would be to use `Map<property-name,property>` like this:

```java
Map<String, KProperty<?>> props = new HashMap<>();

        kObject.getProperties().iterator()
                .forEachRemaining(property -> props.put(property.getName(), property));

        props.forEach((propertyName, property) -> {
            try {
                this.build(toProtobufFormat(propertyName),
                           property.getValue(),
                           protoStreamWriter);
            } catch (IOException e) {
                logger.error("error",
                             e);
            }
        });

        if (kObject.fullText()) {
            protoStreamWriter.writeString(MetaObject.META_OBJECT_FULL_TEXT,
                                          props.values().stream()
                                                  .filter(kProperty -> kProperty.isSearchable() && !(kProperty.getValue() instanceof Boolean))
                                                  .map(kProperty -> String.valueOf(kProperty.getValue()).toLowerCase())
                                                  .collect(joining("\n")));
        }
```